src/lib/optimizer/strategy/semi_join_reduction_rule.cpp

```diff
@@ -0,0 +1,140 @@
+      // which can deal with both sides.
+      const auto reduce_if_beneficial = [&](const auto side_of_join) {
+        // In this block, "left" refers to the left input of the (potential) semi join reduction, i.e., the table that
+        // is getting filtered and "right" refers to the table that the left side is filtered by. As such, right_input
+        // holds the right side of the semi join reduction, i.e., the list of values that the left side is filtered on.
+        // This is independent on the side of the original join that we are operating on, which is called side_of_join.
+        auto left_cardinality = estimator->estimate_cardinality(join_node->input(side_of_join));
+        auto right_input = join_node->input(opposite_side(side_of_join));
+        auto right_cardinality = estimator->estimate_cardinality(right_input);
+
+        const auto semi_join_reduction_node = JoinNode::make(JoinMode::Semi, predicate_expression);
+        semi_join_reduction_node->comment = "Semi Reduction";
+        lqp_insert_node(join_node, side_of_join, semi_join_reduction_node);
+
+        semi_join_reduction_node->set_right_input(right_input);
+
+        const auto reduction_input_cardinality =
+            estimator->estimate_cardinality(semi_join_reduction_node->left_input());
+        const auto reduction_output_cardinality = estimator->estimate_cardinality(semi_join_reduction_node);
+
+        semi_join_reduction_node->set_right_input(nullptr);
+        lqp_remove_node(semi_join_reduction_node);
+
+        if (reduction_input_cardinality == 0 ||
+            reduction_output_cardinality / reduction_input_cardinality > MINIMUM_SELECTIVITY) {
+          return;
+        }
+
+        // For `t1 JOIN t2 on t1.a = t2.a` where a semi join reduction is supposed to be added to the t1, `t1.a = t2.a`
+        // is the predicate_expression. The values from t1 are supposed to be filtered by looking at t1.a, which is
+        // called the left_expression.
+        const auto& left_expression = expression_evaluable_on_lqp(predicate_expression->left_operand(), *right_input)
+                                          ? predicate_expression->left_operand()
+                                          : predicate_expression->right_operand();
+        DebugAssert(!expression_evaluable_on_lqp(left_expression, *join_node->input(side_of_join)),
+                    "Expected filtered expression to be uniquely evaluable on one side of the join");
+
+        // Currently, the right input of the semi join reduction is opposite_side(side_of_join). However, on that side,
+        // multiple joins have been added that increase the cardinality. We walk the LQP starting with right_input
+        // taking those paths where lext_expression continues to be evaluable.
+        while (true) {
+          if (right_input->type != LQPNodeType::Join) {
+            break;
+          }
+          // No matter the join type, we will never see values that were not there before
```

Comment:
> Independent of the join type, we can use either side of the join's input as a candidate for the reducer. This is because the join might only remove values from the reducer's list of values, but does not add any new values. As such, each value in the output of the join will also be found in the input.

Better?

Reply:
"This is because the join" >>  "This is because the added semi join" to distinguish it from the `right_input` join.